p2p/gossip/protocol.go

```diff
@@ -271,38 +286,38 @@ func (prot *Protocol) processMessage(h hash, msg *pb.ProtocolMessage) error {
 	return nil
 }
 
-func (prot *Protocol) handleRelayMessage(msgB []byte) error {
+func (prot *Protocol) handleRelayMessage(msgB []byte) {
 	hash := calcHash(msgB)
+	msg := &pb.ProtocolMessage{}
+	err := proto.Unmarshal(msgB, msg)
+	if err != nil {
+		prot.Log.Error("failed to unmarshal when handling relay message, err %v", err)
+		return
+	}
 
-	// in case the message was received through the relay channel we need to remove the Gossip layer and hand the payload for the next protocol to process
-	if prot.isOldMessage(hash) {
+	// in case the message was received through the relay channel we need to remove the Gossip layer and hand the
+	// payload for the next protocol to process
+	isOld, isInvalid := prot.markAndValidateMessage(hash, msg)
+	if isInvalid {
+		// todo : - have some more metrics for termination
+		prot.Log.Info("got invalid message, hash %d, isOld %v", hash, isOld)
+		return // not propagating invalid messages
+	}
+	if isOld {
 		// todo : - have some more metrics for termination
 		// todo	: - maybe tell the peer weg ot this message already?
-		prot.Log.Debug("got old message, hash %d", hash)
-	} else {
-
-		msg := &pb.ProtocolMessage{}
-		err := proto.Unmarshal(msgB, msg)
-		if err != nil {
-			prot.Log.Error("failed to unmarshal when handling relay message, err %v", err)
-			return err
-		}
+		prot.Log.Debug("got old message, hash %d, isInvalid %v", hash, isInvalid)
 
-		err = prot.validateMessage(msg)
-		if err != nil {
-			prot.Log.Error("failed to validate message when handling relay message, err %v", err)
-			return err
-		}
-
-		err = prot.processMessage(hash, msg)
+	} else {
+		//todo - processMessage is non-blocking, need to check validation and not propagate invalid messages
+		err = prot.processMessage(msg)
 		if err != nil {
-			return err
+			return
```

Comment:
Check if you want to log on that return.

Reply:
This is intentionally, in the next PR I'll change the way we pass messages to protocol handlers